### ISSUE TITLE
feature request: patch validator

### ISSUE THREAD

**Issue by joshua-mcginnis** [Starcounter-Jack/JSON-Patch#29]
Thanks for this project. 

A nice feature would be to validate that a patch document is a valid patch doc. This would allow users to pre-validate that the client sent a valid patch document before the server uses it.

Something like:
`jsonpatch.isValid(patch); //true = valid, false = invalid`

Or you could return errors, but that might be difficult. Ultimately, if a client sends a bad patch, we're just going to return an http bad request.

**Comment by warpech** [comment]
I have pushed to `master` my take on this, which is a generalised version of changes proposed by @joshua-mcginnis, @sonnyp and others.

Please share your thoughts. Below is the description of the new method (excerpt from README.md)

- - - - - - - - 

Validating a sequence of patches:
```js
var obj = {user: {firstName: "Albert"}};
var patches = [{op: "replace", path: "/user/firstName", value: "Albert"}, {op: "replace", path: "/user/lastName", value: "Einstein"}];
var errors = jsonpatch.validate(patches, obj);
if (errors.length == 0) {
 //there are no errors!
}
else {
  for (var i=0; i<errors.length; i++) {
    if (!errors[i]) {
      console.log("Valid patch at index", i, patches[i]);
    }
    else {
      console.error("Invalid patch at index", i, errors[i], patches[i]);
    }
  }
}
```

- - - - - - - -

#### jsonpatch.validate (`patches` Array, `tree` Object (optional)) : `errors` Array

Available in *json-patch-duplex.js*

Validates a sequence of operations. If `tree` parameter is provided, the sequence is additionally validated against the object tree.

If there are errors, returns an array with error codes located at the array index of the operation. If there are no errors, returns an empty array (length 0).

Possible errors:

Error code                    | Description
------------------------------|------------
SEQUENCE_NOT_AN_ARRAY         | Sequence of operations is not an array
OPERATION_NOT_AN_OBJECT       | Operation is not an object
OPERATION_OP_INVALID          | Operation `op` property is not one of operations defined in RFC-6902
OPERATION_PATH_INVALID        | Operation `path` property is not a string
OPERATION_FROM_REQUIRED       | Operation `from` property is not present (applicable in `move` and `copy` operations)
OPERATION_VALUE_REQUIRED      | Operation `value` property is not present (applicable in `add`, `replace` and `test` operations)
OPERATION_PATH_ALREADY_EXISTS | Cannot perform an `add` operation at a path that already exists
OPERATION_PATH_UNRESOLVABLE   | Cannot perform a `replace`, `remove`, `move` or `copy` operation at a path that doesn't exist

**Comment by sonnyp** [comment]
@warpech cool.
However I think the initial multiple option from my PR remains useful. User might not be interested in errors and might want break validation after first error.

**Comment by warpech** [comment]
@sonnyp, you are right about the `add` operation:

> If the target location specifies an object member that does exist,
      that member's value is replaced.

I will fix that.

My motivation for removing the `multiple` option was to have only one behaviour that suits all needs. If you just want to know if there are no errors, check `errors.length === 0`. Otherwise you can iterate over the array to use the detailed information about what and where is the error. I think providing a wider range of options would make it both harder to learn (for users) and maintain (for me).

Thanks for mentioning git rebase, of course I know it but did not want to mess with the merge. My changes are in these commits: 
- https://github.com/Starcounter-Jack/Fast-JSON-Patch/commit/46162f49070793e28776d18335fdad2c248f09ee (changes over your code)
- https://github.com/Starcounter-Jack/Fast-JSON-Patch/commit/55132946d72166f532d373453a1e535ff46f7284 (validate the patch against an object tree)

**Comment by sonnyp** [comment]
@warpech the use case is to stop validation after first error, so removing multiple doesn't suit my needs.
What's wrong with returning an array of length 1 if multiple is set to false?

"both harder to learn " just a flag
"and maintain (for me)" how hard is it to maintain```if (options.multiple === false) ... break```; plus I wrote tests

**Comment by warpech** [comment]
I appreciate your contribution so I want to reflect your needs. Could you describe your use case a little bit more? 

For me, invalid patches are rare and fatal errors, so the efficiency of returning just the first error is irrelevant.

**Comment by sonnyp** [comment]
If you accept incoming patches from untrusted source it allow to prevent process from blocking if a large faulty patch is being validated.

**Comment by warpech** [comment]
I start to get the point. So I guess you are using it on server? Is it in Redbooth?

I think we can:
- add a `stopOnFirstError` boolean parameter
- expose `_validate` method so you can overwrite it if you want to add custom validations (I find it useful for myself)

**Comment by sonnyp** [comment]
@warpech we don't currently use it at redbooth. Yes I use it server side.
Sounds good, thanks. Would you like me to do the PR?

**Comment by warpech** [comment]
No worries, I will do it and let you do the testing :)

**Comment by warpech** [comment]
@sonnyp Pushed. Would you take a look at the commit https://github.com/Starcounter-Jack/Fast-JSON-Patch/commit/b64474b889ced52f8b64b477e0e143448d0a48f8 and share your thoughts?

**Comment by warpech** [comment]
I have reworked the validation. The result is a faster and simpler API.

Summary of changes:
- `jsonpatch.validate` returns and error object now
- `jsonpatch.apply` with third parameter set to `true` throws an error if it encounters an invalid patch

This should please @sonnyp - only the first error is returned now.

See the changes at branch https://github.com/Starcounter-Jack/JSON-Patch/tree/issue-29-simplified
(last 2 commits)

The intention is to release it soon as version 0.6.0

**Comment by warpech** [comment]
Merged to master

**Comment by warpech** [comment]
Released in 0.5.1.

See https://github.com/Starcounter-Jack/JSON-Patch/blob/master/README.md for API and tests for examples

**Comment by gdibble** [comment]
:+1:
